Disable POSIX mode on Windows for cmd.run

opdude opened the pull request:
There was talk about making the string-encode changes more generic and in a single location (#20619), here is my attempt at doing that. It seems we are running commands on windows as if it was POSIX compliant AFAIK it isn't. 

Note: This may break other commands that are trying to use POSIX compliancy on Windows, I have tested quite a few different states but not all :)

opdude committed:
Disable POSIX mode on Windows for cmd.run

- This is a more general fix for the string-encode changes, and therefore I have undone those changes here

opdude commented:
I'll follow up on these test failures tomorrow :)

jfindlay commented on salt/modules/pip.py:
```diff
@@ -127,7 +125,8 @@ def _process_requirements(requirements, cmd, saltenv, user, no_chown):
                 )
                 __salt__['file.chown'](treq, user, None)
                 cleanup_requirements.append(treq)
-            cmd.append('--requirement={0!r}'.format(treq or requirement))
```
This change is causing the pip unit test failures.  We need to test whether pip will accept --requirements='requirements.txt' as a valid argument (instead of --requirements=requirements.txt).

opdude committed:
Fixed pip after POSIX changes

- Encased the requirements file in quotes to allow for spaces in the path
- Moved to using python_shell=True, so that Windows works correctly when you have spacing in the requirements path
- Updated unit tests to reflect the above changes

jfindlay commented:
@opdude, setting `python_shell=True` for commands that accept user input introduces a shell injection vulnerability.

opdude commented:
@jfindlay That is true, I was unable to work out a way to get windows to like spaces in a requirements path without throwing it into the python_shell.

If anyone has any bright ideas I am open to that. I could also just revert the pip changes knowing that it won't work if we change to disabling POSIX on windows

jfindlay commented:
@opdude, sorry that there hasn't been any activity on this for a while, but now that saltconf is over I hope to figure out a way to resolve the `python_shell` issues this week.  Let me know if you've made any progress, thanks.

opdude commented:
@jfindlay No worries I assumed you'd all be busy. No progress on this so far, I hope to take another look at this at some point soon.

jfindlay commented:
@opdude, I've looked this over a few days ago and I have some questions:
- I'm no expert on POSIX, but I am unsure by what you mean by disabling POSIX mode on windows.
- Why does sending cmd.* commands through the shell on windows work properly for encoding, specifically preserving path separators, when bypassing the shell does not?  We cannot enable shell processing by default because of shell injection vulnerabilities, but I am guessing that another modification to cmdmod._run that is specific to windows encoding could solve many windows path separator problems.

thatch45 commented:
Since I think we are going to need to take another direction here I am going to close this one out
